sui-execution/latest/sui-move-natives/src/eth_state_proof.rs

```diff
@@ -47,7 +47,48 @@ pub fn verify_message_proof(
     _ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    todo!();
+    // Load the cost parameters from the protocol config
+    let verify_message_proof_cost_params = &context
+        .extensions()
+        .get::<NativesCostTable>()
+        .eth_state_proof
+        .clone();
+
+    // Charge the base cost for this operation.
+    native_charge_gas_early_exit!(
+        context,
+        verify_message_proof_cost_params.verify_message_proof_cost_base
+    );
+
+    let cost = context.gas_used();
+    let proof = pop_arg!(args, Vector).to_vec_u8()?;
+
+    let Ok(proof) = bcs::from_bytes::<ProofResponse>(proof.as_slice()) else {
+        return Ok(NativeResult::err(
+            cost,
+            StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT.into(),
+        ));
+    };
+
+    // Verify the account proof against the state root.
+    let is_valid_account_proof = verify_proof(
+        &proof.account_proof.proof.as_slice(),
+        &proof.account_proof.root,
+        &proof.account_proof.path,
+        &proof.account_proof.value,
```

Comment:
Here and everywhere @yuvalelmo :
The entire idea of verifying proofs, is that you verify against some statement and then you trust *that same statement* for something. Even the notion of bundling up proofs & statements together in one struct is already "eye-catching" (there are situations in more complex protocols where this is applicable, but generally this is dangerous unless you have a good reason to do so.)

The correct way to design it is to call the verification function with the proof and *statements from the verifier* as opposed to the way it is currently done where the statements come from the prover.

This is a recurring theme in every place we would verify any kind of proof. A good example of where it was done correctly by @yaelAbergel @ItayLevyOfficial is in #123. There, they first [constructed the statement](https://github.com/dwallet-labs/dwallet-network/pull/123/files/e3e7e8963c1181ad84bb8e3b42288f6816e34468#r1705216946) from the `public_key_share` that [comes from the dkg output](https://github.com/dwallet-labs/dwallet-network/blob/e7dd19fbf55816cc567bd8e0f3ad726fc092ed3b/sui-execution/latest/sui-move-natives/src/crypto/twopc_mpc.rs#L96). Then, they [verified the proof](https://github.com/dwallet-labs/dwallet-network/blob/e7dd19fbf55816cc567bd8e0f3ad726fc092ed3b/crates/signature-mpc/src/twopc_mpc_protocols/encrypt_user_share.rs#L227) against that statement, which also includes the `encrypted_user_share` the prover sent. Constructed this way, the proof verification assures the verifier that a relationship exists between the two statements: the public key share and the encrypted user share, thus proving a correct encryption. This is done because the verifier supplied the public key share that it took from a previously known-good state - the DKG output.

Here, verifying against a prover-supplied statements (i.e. `path`, `value`, and `root`) will yield nothing. The correct way is for the verifier to construct these values so that when the proof verifies, it assures them of their validity. 

Actually, a proof usually proves a relationship: so like the encrypted user share proof proves that the encryption is of the secret key share that matches the public key share, here the proof proves a relationship between `path`,`value` and `root`: i.e. that the Merkle tree root of Ethereum state in `root` (that has to be previously verify against the lightclient) has `value` in `path`.

So you must:
1. remove `root`, `path` and `value` from the proof structure
2. take `root` from the `LatestEthereumState` move object (that was previously verified in lightclient)
3. construct `path` from the `message` and `dwallet id` (and `slot`)
4. use `true` for `value`